File: P2PCamera/Codes/QRViewController.m
//
//  QRViewController.m
//  QRWeiXinDemo
//
//  Created by lovelydd on 15/4/25.
//  Copyright (c) 2015年 lovelydd. All rights reserved.
//

#import "QRViewController.h"
#import "BasicDefinetion.h"

#import <AVFoundation/AVFoundation.h>
#import "QRView.h"
@interface QRViewController ()<AVCaptureMetadataOutputObjectsDelegate,QRViewDelegate>

@property (strong, nonatomic) AVCaptureDevice               * device;
@property (strong, nonatomic) AVCaptureDeviceInput          * input;
@property (strong, nonatomic) AVCaptureMetadataOutput       * output;
@property (strong, nonatomic) AVCaptureSession              * session;
@property (strong, nonatomic) AVCaptureVideoPreviewLayer    * preview;

@end

@implementation QRViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setupUI];
    // Do any additional setup after loading the view.
    
    _device = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];
    
    // Input
    _input = [AVCaptureDeviceInput deviceInputWithDevice:self.device error:nil];
    
    // Output
    _output = [[AVCaptureMetadataOutput alloc]init];
    [_output setMetadataObjectsDelegate:self queue:dispatch_get_main_queue()];
    
    // Session
    _session = [[AVCaptureSession alloc]init];
    [_session setSessionPreset:AVCaptureSessionPresetHigh];
    if ([_session canAddInput:self.input])
    {
        [_session addInput:self.input];
    }
    
    if ([_session canAddOutput:self.output])
    {
        [_session addOutput:self.output];
    }
    
    // 条码类型 AVMetadataObjectTypeQRCode
//    _output.metadataObjectTypes =@[AVMetadataObjectTypeQRCode];
    
    //增加条形码扫描
    _output.metadataObjectTypes = @[AVMetadataObjectTypeEAN13Code,
                                    AVMetadataObjectTypeEAN8Code,
                                    AVMetadataObjectTypeCode128Code,
                                    AVMetadataObjectTypeQRCode];
    
    // Preview
    _preview =[AVCaptureVideoPreviewLayer layerWithSession:_session];
    _preview.videoGravity =AVLayerVideoGravityResize;
    _preview.frame =self.view.layer.bounds;
    [self.view.layer insertSublayer:_preview atIndex:0];
    
    [_session startRunning];
    
    CGRect screenRect = [UIScreen mainScreen].bounds;
    QRView *qrRectView = [[QRView alloc] initWithFrame:screenRect];
    qrRectView.transparentArea = CGSizeMake(200*AUTO_WIDTH , 200*AUTO_HEIGHT);
    qrRectView.backgroundColor = [UIColor clearColor];
    qrRectView.center = CGPointMake(self.view.frame.size.width / 2, self.view.frame.size.height / 2);
    qrRectView.delegate = self;
    [self.view addSubview:qrRectView];
    
//    UIButton *jumpBtn = [UIButton buttonWithType:UIButtonTypeCustom];
//    jumpBtn.frame = CGRectMake(SCREEN_WIDTH-70*AUTO_WIDTH, 84*AUTO_HEIGHT, 50*AUTO_WIDTH, 50*AUTO_HEIGHT);
//    [jumpBtn setTitle:@"跳过" forState:UIControlStateNormal];
//    [jumpBtn addTarget:self action:@selector(pop:) forControlEvents:UIControlEventTouchUpInside];
//    [self.view addSubview:jumpBtn];
    
    //修正扫描区域
    CGFloat screenHeight = self.view.frame.size.height;
    CGFloat screenWidth = self.view.frame.size.width;
    CGRect cropRect = CGRectMake((screenWidth - qrRectView.transparentArea.width) / 2,
                                 (screenHeight - qrRectView.transparentArea.height) / 2,
                                 qrRectView.transparentArea.width,
                                 qrRectView.transparentArea.height);

    [_output setRectOfInterest:CGRectMake(cropRect.origin.y / screenHeight,
                                          cropRect.origin.x / screenWidth,
                                          cropRect.size.height / screenHeight,
                                          cropRect.size.width / screenWidth)];

}

- (void)setupUI
{
    [self setMyNavBar];
    self.titleLabel.text = @"扫描";
    [[NSNotificationCenter defaultCenter]addObserver:self selector:@selector(cancel:) name:@"input" object:nil];
}

- (void)cancel:(NSNotification *)notify{
    [self.navigationController popViewControllerAnimated:YES];
}

#pragma mark QRViewDelegate
-(void)scanTypeConfig:(QRItem *)item {
    
    if (item.type == QRItemTypeQRCode) {
        _output.metadataObjectTypes =@[AVMetadataObjectTypeQRCode];
        
    } else if (item.type == QRItemTypeOther) {
        
        _output.metadataObjectTypes = @[AVMetadataObjectTypeEAN13Code,
                                        AVMetadataObjectTypeEAN8Code,
                                        AVMetadataObjectTypeCode128Code,
                                        AVMetadataObjectTypeQRCode];
    }
}
#pragma mark AVCaptureMetadataOutputObjectsDelegate
- (void)captureOutput:(AVCaptureOutput *)captureOutput didOutputMetadataObjects:(NSArray *)metadataObjects fromConnection:(AVCaptureConnection *)connection
{
    NSString *stringValue;
    if ([metadataObjects count] >0)
    {
        //停止扫描
        [_session stopRunning];
        AVMetadataMachineReadableCodeObject * metadataObject = [metadataObjects objectAtIndex:0];
        stringValue = metadataObject.stringValue;
    }
    [[NSUserDefaults standardUserDefaults]setValue:stringValue forKey:@"seriesNumber"];
    [[NSUserDefaults standardUserDefaults]synchronize];
    [[NSNotificationCenter defaultCenter]postNotificationName:@"scanequipmentOK" object:nil];
    
    if (self.qrUrlBlock) {
        self.qrUrlBlock(stringValue);
    }
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    
}





@end

File: P2PCamera/Codes/Main/QRViewController.swift
//
//  QRViewController.swift
//  P2PCamera
//
//  Created by mac on 16/8/9.
//  Copyright © 2016年 Lu. All rights reserved.
//

import UIKit

class QRViewController: UIViewController {

    var ssid:String!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.view.backgroundColor = UIColor.whiteColor()
        
        let image = UIImageView(frame: CGRectMake(40, 64+(SH-64-(SW-80))/3, SW-80, SW-80))
        image.image = QRCodeGenerator.qrImageForString(self.ssid, imageSize: SW-80)
        self.view.addSubview(image)
        
        let button = UIButton(frame: CGRectMake(12,20,44,44))
        button.setBackgroundImage(UIImage(named: "backBtn"), forState: UIControlState.Normal)
        button.addTarget(self, action: #selector(self.popAction), forControlEvents: UIControlEvents.TouchUpInside)
        self.view.addSubview(button)
        
        let line = UIView(frame: CGRectMake(0,64,SW,1))
        line.backgroundColor = UIColor.lightGrayColor()
        self.view.addSubview(line)
        
        let titleLabel = UILabel(frame: CGRectMake(0,20,SW,44))
        titleLabel.font = UIFont.systemFontOfSize(15)
        titleLabel.text = "二维码WiFi"
        titleLabel.textColor = UIColor.blackColor()
        self.view.addSubview(titleLabel)
        titleLabel.textAlignment = NSTextAlignment.Center
        self.showBackButton()
    }
    
    func popAction() {
        self.navigationController?.popViewControllerAnimated(true)
    }

}

extension UIViewController {
    func showBackButton() {
        let button = UIButton (frame: CGRectMake(12,SH-12-40,40,40))
        button.setImage(UIImage(named: "backBtn"), forState: UIControlState.Normal)
        button.addTarget(self, action: #selector(UIViewController.popBtnAction), forControlEvents: UIControlEvents.TouchUpInside)
        self.view.addSubview(button)
    }
    
    func popBtnAction() {
        self.navigationController?.popViewControllerAnimated(true)
    }
    
    func showNaviBackButton() {
        let btn = UIBarButtonItem(image: UIImage(named: "backBtn"), style: UIBarButtonItemStyle.Done, target: self, action: #selector(UIViewController.popBtnAction))
            
        self.navigationItem.backBarButtonItem = btn
    }
}
